### PR TITLE
[SPARK-31765][WEBUI][test-maven] Upgrade HtmlUnit >= 2.37.0

### PR DESCRIPTION
**Pull request by sarutak**
### What changes were proposed in this pull request?

This PR upgrades HtmlUnit.
Selenium and Jetty also upgraded because of dependency.
### Why are the changes needed?

Recently, a security issue which affects HtmlUnit is reported.
https://nvd.nist.gov/vuln/detail/CVE-2020-5529
According to the report, arbitrary code can be run by malicious users.
HtmlUnit is used for test so the impact might not be large but it's better to upgrade it just in case.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Existing testcases.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/ui/JettyUtils.scala [hunk 1]
```diff
@@ -500,7 +501,11 @@ private[spark] case class ServerInfo(
     threadPool match {
       case pool: QueuedThreadPool =>
         // Workaround for SPARK-30385 to avoid Jetty's acceptor thread shrink.
-        pool.setIdleTimeout(0)
+        // As of Jetty 9.4.21, the implementation of
+        // QueuedThreadPoolExecutor#setIdleTimeout is changed and IllegalStateException
+        // will be thrown if we try to set idle timeout after the server has started.
+        // But this workaround works for Jetty 9.4.28 by ignoring the exception.
+        Try(pool.setIdleTimeout(0))
```
Comment (sarutak) [inline core/src/main/scala/org/apache/spark/ui/JettyUtils.scala:508]: I noticed that as of Jetty 9.4.21, the implementation of `QueuedThreadPool`.
When `QueuedThreadPool#setIdleTimeout` is called, `QueuedThreadPool#_idleTimeout` is set to the new timeout as well as [`ReservedThreadExecutor#_idleTimeout`](https://github.com/eclipse/jetty.project/blob/jetty-9.4.28.v20200408/jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java#L312).

If `ReservedThreadExecutor#_idleTImeout` is tried to be set after Jetty has started, [`IllegalStateException` will be thrown](https://github.com/eclipse/jetty.project/blob/jetty-9.4.28.v20200408/jetty-util/src/main/java/org/eclipse/jetty/util/thread/ReservedThreadExecutor.java#L149).

But `QueuedThreadPool#_idleTimeout` will be set so this workaround will still work by ignoring the exception.

**Comment by sarutak** [comment]
Ah, I remember that some tests will fail because the latest HtmlUnit can't properly interpret JavaScript code used in those tests.
So PR should follow #28578 .

**Comment by srowen** [comment]
Looks OK pending tests

**Comment by HyukjinKwon** [comment]
cc @gengliangwang too FYI. seems okay to me too

File: core/src/test/scala/org/apache/spark/ui/UISeleniumSuite.scala [hunk 2]
```diff
@@ -784,6 +784,7 @@ class UISeleniumSuite extends SparkFunSuite with WebBrowser with Matchers with B
 
       eventually(timeout(10.seconds), interval(50.milliseconds)) {
         goToUi(sc, "/jobs")
+
```
Comment (sarutak) [inline core/src/test/scala/org/apache/spark/ui/UISeleniumSuite.scala:787]: I'll remove this extra empty line.

**Review by gengliangwang — APPROVED** [review]
LGTM, thanks for the work!

**Comment by gengliangwang** [comment]
merging to master

**Comment by sarutak** [comment]
@gengliangwang This PR doesn't pass test.
As I mentioned [here](https://github.com/apache/spark/pull/28585#issuecomment-631255845), #28578 should be merged first.

**Comment by gengliangwang** [comment]
@sarutak sorry about that, let me do the revert

**Comment by srowen** [comment]
I merged https://github.com/apache/spark/pull/28578 ; this may need a rebase

**Comment by sarutak** [comment]
According to the last failure, I noticed that real headless browser support is needed for HistoryServer tests too.
I've open a PR for that.
#28622

**Comment by srowen** [comment]
Jenkins retest this please

**Comment by srowen** [comment]
Merged to master

**Comment by dongjoon-hyun** [comment]
Hi, All.
This seems to break all Maven test. Could you take a look at them?

**Comment by HyukjinKwon** [comment]
Let's revert for now since it broke the tests.

**Comment by HyukjinKwon** [comment]
Reverted manually at baafd4386c89d61c5eb600554df6315466e95fa8 for now. Let's just redo it in a separate PR since it will need to test it out in different combinations of profiles which will likely take a while.

**Comment by dongjoon-hyun** [comment]
+1 for that. Thanks!

**Comment by HyukjinKwon** [comment]
Thanks @dongjoon-hyun for monitoring this.

**Comment by sarutak** [comment]
Thanks @dongjoon-hyun , @HyukjinKwon .
I don't know why sbt pass the tests but I'll investigate it the reason anyway.

**Comment by dongjoon-hyun** [comment]
Thanks, @sarutak .

**Comment by sarutak** [comment]
@dongjoon-hyun @HyukjinKwon The reason of the failure seems that `bytebuddy` which upgraded-selenium depends on is much older than `mockito` depends on.
Let's wait the result of CI.

**Comment by dongjoon-hyun** [comment]
Sounds reasonable. Thanks!

**Comment by sarutak** [comment]
retest this please.

**Comment by sarutak** [comment]
retest this please.

**Comment by sarutak** [comment]
retest this please.

**Comment by sarutak** [comment]
By excluding old version of `bytebuddy` , Maven tests seem to pass.

**Comment by srowen** [comment]
Heh, do we think it's safe to merge again?
An exclusion is probably fine, but now it depends on something it needs only transitively. We could also manage the version explicitly. But, this maybe fine now.

**Comment by sarutak** [comment]
The errors were reproduced in PR builder and they are fixed by excluding the old `bytebuddy` so I believe it's safe to merge.
What do you think, @dongjoon-hyun ?

**Comment by srowen** [comment]
I'll merge today if there are no objections

**Comment by srowen** [comment]
Trying again - merged to master

**Comment by dongjoon-hyun** [comment]
According to @sarutak , this broke `org.apache.spark.deploy.history.ChromeUIHistoryServerSuite` due to Guava dependency.
- https://github.com/apache/spark/pull/29827#discussion_r492654755

**Comment by srowen** [comment]
Shoot. If this is non-essential, I think we can just revert it.
Or .. just lose this test? let me comment on the other PR.

**Comment by dongjoon-hyun** [comment]
Thanks, @srowen ~

**Pull request closed without merging**